# Load libraries

In [149]:
import pandas as pd 
import numpy as np
import re

# Load datasets

In [150]:
df1 = pd.read_csv("data/ALLARMI.csv")
df2 = pd.read_csv("data/TIPOLOGIA_VIAGGIATORE.csv")

In [151]:
df1

,OCCORRENZE,AREOPORTO_ARRIVO,AREOPORTO_PARTENZA,ANNO_PARTENZA,MESE_PARTENZA,DATA_PARTENZA,DESCR_AEREOPORTO_ARR,DESCR_AEREOPORTO_PART,CITTA_ARR,CITTA_PARTENZA,...,ZONA,TOT,MOTIVO_ALLARME,note_operatore,flag_rischio,Paese Partenza,CODICE PAESE ARR,3zona,paese%arr,tot voli
0,Voli con Allarmi,FCO,IST,2024,01,2024-01-30 09:15:00,Fiumicino,Havalimani,Roma,Istanbul,...,5,1,Manuale,NaN,NaN,Turchia,ITA,5,Italia,1
1,Viaggiatori con Allarmi,CIA,STN,2024,02,2024-02-03 13:15:00,Ciampino,Stansted,Roma,Londra,...,5,5,Manuale,NaN,NaN,Regno Unito,ITA,5,Italia,5
2,Viaggiatori entrati nel Sistema,FCO,LHR,2024,01,2024-01-15 08:45:00,Fiumicino,London Heathrow,Roma,Londra,...,5,110,TSC,NaN,NaN,Regno Unito,ITA,5,Italia,110
3,Voli con Allarmi,MXP,LHR,2024,02,2024-02-02 08:40:00,Malpensa,London Heathrow,Milano,Londra,...,2,1,SDI,NaN,NaN,Regno Unito,ITA,2,Italia,1
4,Viaggiatori con Allarmi,PSA,BRS,2024,02,2024-02-16 12:50:00,Galileo Galilei,Bristol,Pisa,Bristol,...,8,2,INTERPOL,NaN,NaN,Regno Unito,ITA,8,Italia,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5075,Nulla a procedere SDI,MXP,LTN,2023,01,2024-01-19 17:30:00,Malpensa,London Luton,Milano,Londra,...,2,5,TSC,NaN,NaN,Regno Unito,ITA,2,Italia,5
5076,Voli disponibili in ingresso al Sistema,BLQ,RAK,2024,02,FEB 06 2024,Guglielmo Marconi,Menara,Bologna,Marrakech,...,8,1,TSC,NaN,NaN,Marocco,ITA,8,Italia,1
5077,Voli con Allarmi,FCO,DOH,2024,01,2024-01-12 15:35:00,Fiumicino,Hamad International,Roma,Doha,...,5,1,NSIS,NaN,NaN,Qatar,ITA,5,Italia,1
5078,"Voli solo visualizzati, ma NON investigati",LIN,LHR,2024,01,2024-01-13 11:50:00,Linate,London Heathrow,Milano,Londra,...,2,1,TSC,NaN,NaN,Regno Unito,ITA,2,Italia,1


In [152]:
df2

,NAZIONALITA,AREOPORTO_ARRIVO,AREOPORTO_PARTENZA,ANNO_PARTENZA,MESE_PARTENZA,GIORNO_PARTENZA,DATA_PARTENZA,DESCR_AEREOPORTO_ARR,DESCR_AEREOPORTO_PART,CITTA_ARR,...,COMPAGNIA_AEREA,NUMERO_VOLO,ESITO_CONTROLLO,note_operatore,codice_rischio,Tipo Documento,FASCIA ETA,3nazionalita,compagnia%aerea,num volo
0,ALB,NAP,DUR,2024,02,13,2024-02-13 07:30:00,Napoli Capodichino,King Shaka International,Napoli,...,Fly Dubai,FZ1681,RESPINTO,NaN,NaN,Passaporto,N.D.,ALB,Fly Dubai,FZ1681
1,NaN,FCO,JFK,2024,01,22,2024-01-22 16:35:00,Fiumicino,John F Kennedy International,Roma,...,ITA Airways,AZ0609,NaN,NaN,NaN,Carta d'identità,18-30,ALB,ITA Airways,AZ0609
2,ALB,TSF,TIA,2024,02,4,2024-02-04 20:10:00,Treviso-Sant'Angelo,Rinas Mother Teresa,Treviso,...,Ryanair DAC,FR8400,SEGNALATO,NaN,NaN,N.D.,31-45,ALB,Ryanair DAC,FR8400
3,AFG,FCO,IST,2024,01,25,2024-01-25 13:05:00,Fiumicino,Havalimani,Roma,...,Turkish Airlines,TK1865,NaN,NaN,NaN,N.D.,61+,AFG,Turkish Airlines,TK1865
4,ALB,BGY,MLE,2024,02,13,FEB 13 2024,Orio al Serio,Male International,Bergamo,...,Fly Dubai,FZ1571,SEGNALATO,NaN,NaN,Permesso di soggiorno,46-60,ALB,Fly Dubai,FZ1571
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5090,ALB,BLQ,TIA,2024,02,15,2024-02-18 09:10:00,Guglielmo Marconi,Rinas Mother Teresa,Bologna,...,Ryanair DAC,FR8399,NaN,NaN,NaN,N.D.,0-17,ALB,Ryanair DAC,FR8399
5091,ALB,FCO,TIA,2024,01,10,2024-01-10 06:10:00,Fiumicino,Rinas Mother Teresa,Roma,...,Wizz Air Malta,W43857,NaN,NaN,NaN,Passaporto,46-60,ALB,Wizz Air Malta,W43857
5092,ALB,FCO,JED,2024,01,31,31/01/2024 22:15:00,Fiumicino,King Abdulaziz International,Roma,...,Wizz Air Malta,w48350,FERMATO,NaN,NaN,N.D.,31-45,ALB,Wizz Air Malta,w48350
5093,ALB,BGY,TIA,2024,01,1,2024-01-01 08:25:00,Orio al Serio,Rinas Mother Teresa,Bergamo,...,Ryanair DAC,FR8350,IN ATTESA,NaN,NaN,Carta d'identità,18-30,ALB,Ryanair DAC,FR8350


# Allarmi Dataset

First general look at the dataset (df1)

In [153]:
df1.shape

(5080, 24)

In [154]:
df1.head(5)

,OCCORRENZE,AREOPORTO_ARRIVO,AREOPORTO_PARTENZA,ANNO_PARTENZA,MESE_PARTENZA,DATA_PARTENZA,DESCR_AEREOPORTO_ARR,DESCR_AEREOPORTO_PART,CITTA_ARR,CITTA_PARTENZA,...,ZONA,TOT,MOTIVO_ALLARME,note_operatore,flag_rischio,Paese Partenza,CODICE PAESE ARR,3zona,paese%arr,tot voli
0,Voli con Allarmi,FCO,IST,2024,01,2024-01-30 09:15:00,Fiumicino,Havalimani,Roma,Istanbul,...,5,1,Manuale,NaN,NaN,Turchia,ITA,5,Italia,1
1,Viaggiatori con Allarmi,CIA,STN,2024,02,2024-02-03 13:15:00,Ciampino,Stansted,Roma,Londra,...,5,5,Manuale,NaN,NaN,Regno Unito,ITA,5,Italia,5
2,Viaggiatori entrati nel Sistema,FCO,LHR,2024,01,2024-01-15 08:45:00,Fiumicino,London Heathrow,Roma,Londra,...,5,110,TSC,NaN,NaN,Regno Unito,ITA,5,Italia,110
3,Voli con Allarmi,MXP,LHR,2024,02,2024-02-02 08:40:00,Malpensa,London Heathrow,Milano,Londra,...,2,1,SDI,NaN,NaN,Regno Unito,ITA,2,Italia,1
4,Viaggiatori con Allarmi,PSA,BRS,2024,02,2024-02-16 12:50:00,Galileo Galilei,Bristol,Pisa,Bristol,...,8,2,INTERPOL,NaN,NaN,Regno Unito,ITA,8,Italia,2


In [155]:
df1.info()

<class 'pandas.DataFrame'>
RangeIndex: 5080 entries, 0 to 5079
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype
---  ------                 --------------  -----
 0   OCCORRENZE             5080 non-null   str  
 1   AREOPORTO_ARRIVO       5080 non-null   str  
 2   AREOPORTO_PARTENZA     5080 non-null   str  
 3   ANNO_PARTENZA          5080 non-null   str  
 4   MESE_PARTENZA          5080 non-null   str  
 5   DATA_PARTENZA          5080 non-null   str  
 6   DESCR_AEREOPORTO_ARR   5080 non-null   str  
 7   DESCR_AEREOPORTO_PART  4971 non-null   str  
 8   CITTA_ARR              5080 non-null   str  
 9   CITTA_PARTENZA         4979 non-null   str  
 10  CODICE_PAESE_ARR       5080 non-null   str  
 11  CODICE_PAESE_PART      5026 non-null   str  
 12  PAESE_ARR              5080 non-null   str  
 13  PAESE_PART             5006 non-null   str  
 14  ZONA                   5080 non-null   str  
 15  TOT                    5080 non-null   str  
 16 

In [ ]:
# Normalizing the dataset, lowercasing and stripping useless cases
df1 = df1.map(lambda x: x.strip().lower() if isinstance(x, str) else x)

# Checking for duplicates and dropping them 
print('Initial duplicates:')
print(df1.duplicated().sum())
df1 = df1.drop_duplicates()
print('Duplicates after cleaning:')
print(df1.duplicated().sum())


Initial duplicates:
50
Duplicates after cleaning:
0


### Occorrenze

In [158]:
print(f"Unique values: {df1['OCCORRENZE'].nunique()}")
print(f"Null values: {df1['OCCORRENZE'].isnull().sum()}")
print()
print(df1['OCCORRENZE'].value_counts(dropna=False).to_string())

Unique values: 30
Null values: 0

OCCORRENZE
viaggiatori entrati nel sistema                 618
voli disponibili in ingresso al sistema         583
voli solo visualizzati, ma non investigati      563
voli investigati (sdi/nsis - interpol - tsc)    540
viaggiatori investigati                         520
allarmi generati da sdi/nsis                    509
voli con allarmi                                505
viaggiatori con allarmi                         463
nulla a procedere sdi                           247
allarmi chiusi                                  243
allarmi rilevanti                               112
errata segnalazione sdi                          25
allarmi generati da bcs                          21
allarmi chiusi con azione (cc.xx)                14
notifica atti/provv                              10
allarmi generati da interpol                      8
allarmi non chiusi                                7
mancato aggiornamento sdi                         6
nulla a procedere n

Grouping categories that are similar to each other

In [159]:
occorrenze_map = {
    'viaggiatori entrati nel sistema'              : 'viaggiatori entrati',
    'voli disponibili in ingresso al sistema'      : 'voli disponibili',
    'voli solo visualizzati, ma non investigati'   : 'voli solo visualizzati',
    'voli investigati (sdi/nsis - interpol - tsc)' : 'voli investigati',
    'viaggiatori investigati'                      : 'viaggiatori investigati',
    'voli con allarmi'                             : 'voli con allarmi',
    'viaggiatori con allarmi'                      : 'viaggiatori con allarmi',
    'allarmi generati da sdi/nsis'                 : 'allarmi generati',
    'allarmi generati da bcs'                      : 'allarmi generati',
    'allarmi generati da interpol'                 : 'allarmi generati',
    'allarmi generati'                             : 'allarmi generati',
    'allarmi chiusi'                               : 'allarmi chiusi',
    'allarmi chiusi con azione (cc.xx)'            : 'allarmi chiusi',
    'allarmi rilevanti'                            : 'allarmi rilevanti',
    'allarmi non chiusi'                           : 'allarmi non chiusi',
    'nulla a procedere sdi'                        : 'esito negativo',
    'nulla a procedere nsis'                       : 'esito negativo',
    'nulla a procedere int'                        : 'esito negativo',
    'errata segnalazione sdi'                      : 'esito negativo',
    'errata segnalazione nsis'                     : 'esito negativo',
    'errata segnalazione bcs'                      : 'esito negativo',
    'respinto/a'                                   : 'esito negativo',
    'mancato aggiornamento sdi'                    : 'mancato aggiornamento',
    'mancato aggiornamento schengen nsis'          : 'mancato aggiornamento',
    'notifica atti/provv'                          : 'altro',
    'inammissibilita schengen - art.24'            : 'altro',
    'altro'                                        : 'altro',
    '???'                                          : np.nan,
    'n/c'                                          : np.nan,
    'ali armatt'                                   : np.nan,
}

df1['OCCORRENZE'] = df1['OCCORRENZE'].map(occorrenze_map)

print(f"\nNull Values: {df1['OCCORRENZE'].isnull().sum()}")
print(f"\nDistribution:")
print(df1['OCCORRENZE'].value_counts(dropna=False).to_string())


Null Values: 3

Distribution:
OCCORRENZE
viaggiatori entrati        618
voli disponibili           583
voli solo visualizzati     563
allarmi generati           543
voli investigati           540
viaggiatori investigati    520
voli con allarmi           505
viaggiatori con allarmi    463
esito negativo             292
allarmi chiusi             257
allarmi rilevanti          112
altro                       16
mancato aggiornamento        8
allarmi non chiusi           7
NaN                          3


### Aeroporto Arrivo

In [160]:
print(f"Unique values: {df1['AREOPORTO_ARRIVO'].nunique()}")
print(f"Null values: {df1['AREOPORTO_ARRIVO'].isnull().sum()}")
print()
print(df1['AREOPORTO_ARRIVO'].value_counts(dropna=False).to_string())

Unique values: 31
Null values: 0

AREOPORTO_ARRIVO
fco    1272
mxp    1185
bgy     464
vce     356
blq     326
nap     202
lin     174
cia     163
trn     151
psa     139
cta     114
vrn     108
tsf      82
flr      68
bri      57
pmo      26
aoi      24
goa      20
peg      16
psr      15
trs      14
rmi      12
suf       9
pmf       7
cag       7
bds       6
cuf       6
ciy       3
bzo       2
olb       1
vbs       1


In [161]:
# ── df1 – AREOPORTO_PARTENZA ─────────────────────────────────────────────────
print(f"Valori unici: {df1['AREOPORTO_PARTENZA'].nunique()}")
print(f"Valori nulli: {df1['AREOPORTO_PARTENZA'].isnull().sum()}")
print()
print(df1['AREOPORTO_PARTENZA'].value_counts(dropna=False).to_string())

Valori unici: 112
Valori nulli: 0

AREOPORTO_PARTENZA
tia    612
stn    502
lgw    401
lhr    387
ist    300
dxb    153
cai    133
saw    132
cmn    130
man    129
doh    105
rak    105
jfk    104
ssh     95
tun     86
beg     81
auh     77
lcy     73
ewr     63
ltn     58
jed     52
gru     51
edi     50
brs     46
amm     46
add     46
tlv     45
rmo     43
bhx     41
pek     35
eze     31
evn     31
spx     30
pvg     30
icn     28
del     27
tpe     25
alg     25
kwi     23
sin     22
ruh     21
mct     20
ema     18
skp     17
mle     17
szx     17
gig     17
atl     16
mia     16
kiv     15
gyd     15
ncl     15
rmf     15
shj     15
wnz     13
bah     13
ika     12
dss     12
yul     12
kut     12
tgd     11
phl     11
mex     10
gla     10
hnd     10
fez      9
lrm      9
hrg      9
sid      9
mba      9
hkg      8
tng      8
dfw      8
prn      8
bqh      8
dwc      8
fab      7
bey      7
hgh      6
yyz      6
lpl      6
tbs      6
iad      6
adb      6
znz      5
nkg      5


### Anno Partenza

In [162]:
print(f"Unique values: {df1['ANNO_PARTENZA'].nunique()}")
print(f"Null values: {df1['ANNO_PARTENZA'].isnull().sum()}")
print()
print(df1['ANNO_PARTENZA'].value_counts(dropna=False).to_string())

Unique values: 5
Null values: 0

ANNO_PARTENZA
2024         4631
2023          256
24             52
anno 2024      49
2024.          42


In [163]:
def normalizza_anno(val):
    if pd.isna(val): return np.nan
    val = str(val).strip()
    match = re.search(r'\b(20\d{2})\b', val)
    if match: return int(match.group(1))
    match = re.search(r'\b(\d{2})\b', val)
    if match: return int('20' + match.group(1))
    return np.nan

df1['ANNO_PARTENZA'] = df1['ANNO_PARTENZA'].apply(normalizza_anno)

print(df1['ANNO_PARTENZA'].value_counts(dropna=False).to_string())

ANNO_PARTENZA
2024    4774
2023     256


### Mese Partenza

In [164]:
print(f"Unique values: {df1['MESE_PARTENZA'].nunique()}")
print(f"Null values: {df1['MESE_PARTENZA'].isnull().sum()}")
print()
print(df1['MESE_PARTENZA'].value_counts(dropna=False).to_string())

Unique values: 14
Null values: 0

MESE_PARTENZA
01     2425
02     2273
gen      80
feb      63
12       26
07       23
06       22
04       21
11       19
09       17
03       16
08       15
05       15
10       15


In [165]:
mesi_map = {"gen": "01", "feb": "02"}
df1["MESE_PARTENZA"] = df1["MESE_PARTENZA"].replace(mesi_map)

print(df1["MESE_PARTENZA"].value_counts(dropna=False).sort_index().to_string())

MESE_PARTENZA
01    2505
02    2336
03      16
04      21
05      15
06      22
07      23
08      15
09      17
10      15
11      19
12      26


### Data Partenza

In [166]:
print(f"Unique values: {df1['DATA_PARTENZA'].nunique()}")
print(f"Null values: {df1['DATA_PARTENZA'].isnull().sum()}")
print()
print(df1['DATA_PARTENZA'].value_counts(dropna=False).to_string())

Unique values: 3810
Null values: 0

DATA_PARTENZA
2024-02-04 06:10:00    6
2024-01-31 18:00:00    6
2024-01-19 08:55:00    5
2024-01-22 13:00:00    5
2024/02/26             5
2024-01-04 11:30:00    5
2024-01-06 08:00:00    5
2024-02-29 16:40:00    5
2024-02-18 06:50:00    5
2024-02-09 12:50:00    4
05.01.2024             4
gen 06 2024            4
feb 18 2024            4
2024-01-30 12:35:00    4
18.02.2024             4
2024-01-22 07:30:00    4
2024-02-19 09:00:00    4
2024-02-10 22:20:00    4
gen 14 2024            4
2024-02-08 10:00:00    4
17-01-24               4
2024-02-18 13:05:00    4
2024-01-01 08:00:00    4
2024-02-02 09:15:00    4
2024-01-05 19:05:00    4
2024-01-27 18:05:00    4
23-02-24               4
2024-01-24 08:00:00    4
2024-02-09 18:05:00    4
2024-01-21 17:50:00    4
2024-02-19 17:55:00    4
2024-01-07 18:00:00    4
2024-01-04 09:25:00    4
2024-02-13 17:55:00    4
2024-02-02 17:20:00    4
2024-02-02 12:50:00    4
2024-02-09 13:20:00    4
2024-02-19 20:30:00    4


In [167]:
mesi_ita = {
    "GEN": "01", "FEB": "02", "MAR": "03", "APR": "04",
    "MAG": "05", "GIU": "06", "LUG": "07", "AGO": "08",
    "SET": "09", "OTT": "10", "NOV": "11", "DIC": "12"
}

def normalizza_data(val):
    val = str(val).strip()

    # "GEN 06 2024" → "2024-01-06 00:00:00"
    for ita, num in mesi_ita.items():
        if val.startswith(ita + " "):
            parts = val.split(" ")
            return f"{parts[2]}-{num}-{parts[1].zfill(2)} 00:00:00"

    # "2024-01-05T12:35:00" → "2024-01-05 12:35:00"
    val = val.replace("T", " ")

    # "2024/02/26" o "2024/02/26 09:00:00" → "2024-02-26 00:00:00"
    if re.match(r"\d{4}/\d{2}/\d{2}", val):
        val = val.replace("/", "-")
        if len(val) == 10:
            val = val + " 00:00:00"

    # "05.01.2024" → "2024-01-05 00:00:00"
    match = re.match(r"(\d{2})\.(\d{2})\.(\d{4})", val)
    if match:
        return f"{match.group(3)}-{match.group(2)}-{match.group(1)} 00:00:00"

    # "04-01-24" → "2024-01-04 00:00:00"
    match = re.match(r"(\d{2})-(\d{2})-(\d{2})$", val)
    if match:
        return f"20{match.group(3)}-{match.group(2)}-{match.group(1)} 00:00:00"

    # "22/01/2024 15:45:00" → "2024-01-22 15:45:00"
    match = re.match(r"(\d{2})/(\d{2})/(\d{4})\s(\d{2}:\d{2}:\d{2})", val)
    if match:
        return f"{match.group(3)}-{match.group(2)}-{match.group(1)} {match.group(4)}"

    return val

df1["DATA_PARTENZA"] = df1["DATA_PARTENZA"].apply(normalizza_data)
df1["DATA_PARTENZA"] = pd.to_datetime(df1["DATA_PARTENZA"], errors="coerce")


print("NaT left:", df1["DATA_PARTENZA"].isnull().sum())
print("Range date:", df1["DATA_PARTENZA"].min(), "→", df1["DATA_PARTENZA"].max())
print("Examples:")
print(df1["DATA_PARTENZA"].head(5).to_string())

NaT left: 129
Range date: 2023-12-31 00:00:00 → 2024-02-29 22:20:00
Examples:
0   2024-01-30 09:15:00
1   2024-02-03 13:15:00
2   2024-01-15 08:45:00
3   2024-02-02 08:40:00
4   2024-02-16 12:50:00


### Descr Aeroporto Arr

In [168]:
print(f"Unique values: {df1['DESCR_AEREOPORTO_ARR'].nunique()}")
print(f"Null values: {df1['DESCR_AEREOPORTO_ARR'].isnull().sum()}")
print()
print(df1['DESCR_AEREOPORTO_ARR'].value_counts(dropna=False).to_string())

Unique values: 31
Null values: 0

DESCR_AEREOPORTO_ARR
fiumicino               1272
malpensa                1185
orio al serio            464
venezia marco polo       356
guglielmo marconi        326
napoli capodichino       202
linate                   174
ciampino                 163
sandro pertini           151
galileo galilei          139
fontanarossa             114
verona villafranca       108
treviso-sant'angelo       82
firenze peretola          68
palese                    57
punta raisi               26
falconara                 24
cristoforo colombo        20
sant egidio               16
liberi                    15
ronchi dei legionari      14
miramare                  12
lamezia terme              9
parma                      7
elmas                      7
papola casale              6
levaldigi                  6
comiso                     3
bolzano                    2
costa smeralda             1
brescia montichiari        1


### Descr Aeroporto Part

In [169]:
print(f"Unique values: {df1['DESCR_AEREOPORTO_PART'].nunique()}")
print(f"Null values: {df1['DESCR_AEREOPORTO_PART'].isnull().sum()}")
print()
print(df1['DESCR_AEREOPORTO_PART'].value_counts(dropna=False).to_string())

Unique values: 119
Null values: 109

DESCR_AEREOPORTO_PART
rinas mother teresa                                                  578
stansted                                                             472
london gatwick                                                       381
london heathrow                                                      364
havalimani                                                           284
dubai international                                                  142
cairo international                                                  126
sabiha gokcen                                                        125
manchester                                                           122
mohamed v                                                            120
NaN                                                                  109
hamad international                                                  102
menara                                                           

The column contained 3 types of issues:

1. **167 rows with extra whitespace** — fixed with `.str.strip()`

2. **83 placeholder values** — values such as `"n.d."`, `"nd"`, `"unknown"`, `"null"`
   that carry no real information but were written as text instead of
   being left empty. Converted to `pd.NA` using `.replace()`.

3. **109 original nulls** — empty cells already present in the dataset.

Points 2 and 3 brought the total null count to 192 (3.8% of rows).
All 192 were **recovered** using the IATA code present in the
`AREOPORTO_PARTENZA` column: through an IATA → full airport name
lookup table, each null was replaced with the correct name in lowercase
(e.g. `"STN"` → `"stansted"`, `"LHR"` → `"london heathrow"`).

**Final result: 0 nulls, 133 unique values, all lowercase.**

In [170]:
# ── Dizionario IATA → nome descrittivo ───────────────────────────────────────

PLACEHOLDER_NULL = {'n.d.', '?', '', '//', '-', 'unknown', 'nd', 'n/a', 'null', 'anu', 'none'}

# Prendo solo le righe dove il nome descrittivo è valido (non placeholder)
df1_validi = df1[~df1['DESCR_AEREOPORTO_PART'].isin(PLACEHOLDER_NULL) & df1['DESCR_AEREOPORTO_PART'].notna()]

iata_to_name = (
    df1_validi
    .groupby('AREOPORTO_PARTENZA')['DESCR_AEREOPORTO_PART']
    .agg(lambda x: x.mode()[0])
    .to_dict()
)

print(f"✅ Dizionario creato: {len(iata_to_name)} aeroporti mappati")

df1['DESCR_AEREOPORTO_PART'] = df1['DESCR_AEREOPORTO_PART'].apply(
    lambda x: np.nan if str(x).strip() in PLACEHOLDER_NULL else x
)

print(f"Nulli dopo sostituzione placeholder: {df1['DESCR_AEREOPORTO_PART'].isnull().sum()}")

df1['DESCR_AEREOPORTO_PART'] = df1.apply(
    lambda row: iata_to_name.get(row['AREOPORTO_PARTENZA'], row['AREOPORTO_PARTENZA'])
    if pd.isna(row['DESCR_AEREOPORTO_PART'])
    else row['DESCR_AEREOPORTO_PART'],
    axis=1
)

print(f"Nulli rimasti: {df1['DESCR_AEREOPORTO_PART'].isnull().sum()}")
print(df1['DESCR_AEREOPORTO_PART'].value_counts(dropna=False).to_string())

✅ Dizionario creato: 111 aeroporti mappati
Nulli dopo sostituzione placeholder: 304
Nulli rimasti: 0
DESCR_AEREOPORTO_PART
rinas mother teresa                                                  612
stansted                                                             502
london gatwick                                                       401
london heathrow                                                      387
havalimani                                                           300
dubai international                                                  153
cairo international                                                  133
sabiha gokcen                                                        132
mohamed v                                                            130
manchester                                                           129
hamad international                                                  105
menara                                                               105
j

### Città Arrivo

In [171]:
print(f"Unique values: {df1['CITTA_ARR'].nunique()}")
print(f"Null values: {df1['CITTA_ARR'].isnull().sum()}")
print()
print(df1['CITTA_ARR'].value_counts(dropna=False).to_string())

Unique values: 29
Null values: 0

CITTA_ARR
roma             1435
milano           1359
bergamo           464
venezia           356
bologna           326
napoli            202
torino            151
pisa              139
catania           114
verona            108
treviso            82
firenze            68
bari               57
palermo            26
ancona             24
genova             20
perugia            16
pescara            15
trieste            14
rimini             12
lamezia terme       9
parma               7
cagliari            7
brindisi            6
cuneo               6
comiso              3
bolzano             2
olbia               1
brescia             1


### Città Partenza

In [172]:
print(f"Unique values: {df1['CITTA_PARTENZA'].nunique()}")
print(f"Null values: {df1['CITTA_PARTENZA'].isnull().sum()}")
print()
print(df1['CITTA_PARTENZA'].value_counts(dropna=False).to_string())

Unique values: 110
Null values: 100

CITTA_PARTENZA
londra               1362
tirana                582
istanbul              408
dubai                 152
cairo                 125
manchester            124
casablanca            120
marrakech             102
NaN                   100
doha                   99
new york               99
sharm el sheikh        88
tunis                  81
belgrado               80
abu dhabi              73
newark                 57
chisinau               55
jeddah                 50
edinburgh              50
san paolo              48
bristol                45
tel aviv yafo          44
amman                  43
addis ababa            41
birmingham             38
pechino                34
n.d.                   31
yerevan                30
buenos aires           29
shanghai               29
seoul                  28
giza                   27
nuova delhi            26
algeri                 25
taipei                 24
?                      23
nd          

In [173]:
# Recover nulls via IATA → city lookup
iata_city = {
    "tia": "tirana",   "stn": "londra",    "lgw": "londra",   "lhr": "londra",
    "lcy": "londra",   "ltn": "londra",    "bqh": "londra",   "fab": "londra",
    "ist": "istanbul", "dxb": "dubai",     "cai": "il cairo", "saw": "istanbul",
    "cmn": "casablanca","doh": "doha",      "rak": "marrakech","jfk": "new york",
    "ssh": "sharm el-sheikh","tun": "tunisi","beg": "belgrado","auh": "abu dhabi",
    "ewr": "new york", "jed": "jeddah",    "edi": "edinburgo","gru": "san paolo",
    "eze": "buenos aires","add": "addis abeba","mex": "citta del messico",
    "pek": "pechino",  "rmo": "chisinau",  "bhx": "birmingham","pvg": "shanghai",
    "hkg": "hong kong","mct": "muscat",    "mle": "male",     "rba": "rabat",
    "evn": "yerevan",  "tbs": "tbilisi",   "ema": "east midlands","ruh": "riad",
    "shj": "sharjah",  "gla": "glasgow",   "icn": "seul",     "rmf": "marsa alam",
    "tng": "tangeri",  "sid": "praia",     "spx": "cairo",    "ika": "teheran",
    "prn": "pristina", "mba": "mombasa",   "szx": "shenzhen", "mia": "miami",
    "tpe": "taipei",   "sin": "singapore", "tlv": "tel aviv", "amm": "amman",
    "brs": "bristol",  "man": "manchester","anu": "saint john's","ncl": "newcastle",
    "del": "nuova delhi","atl": "atlanta", "kwi": "kuwait",   "dfw": "dallas",
    "iad": "washington","yul": "montreal", "yyz": "toronto",  "hnd": "tokyo",
    "bey": "beirut",   "alg": "algeri",    "kiv": "chisinau", "gyd": "baku",
    "lpl": "liverpool","skp": "skopje",    "tgd": "podgorica","phl": "filadelfia",
    "fez": "fez",      "gig": "rio de janeiro","atq": "amritsar","znz": "zanzibar",
    "nkg": "nanchino", "aga": "agadir",    "mji": "misurata", "dmm": "dammam",
    "ckg": "chongqing","bfs": "belfast",   "bos": "boston",   "lxr": "luxor",
    "can": "guangzhou","wnz": "wenzhou",   "hgh": "hangzhou", "bah": "manama",
    "dss": "dakar",    "kut": "kutaisi",   "hrg": "hurghada", "lrm": "al marj",
    "oak": "oakland",  "lax": "los angeles","oxf": "oxford",  "lba": "leeds",
    "ala": "almaty",   "isl": "islamabad", "lyx": "marsa matruh","nht": "londra",
    "mbj": "montego bay","cun": "cancun",  "mru": "port louis","tas": "tashkent",
    "sll": "freetown", "cgo": "zhengzhou", "tfu": "chengdu",  "lyx": "marsa matruh",
}

null_mask = df1["CITTA_PARTENZA"].isnull()
df1.loc[null_mask, "CITTA_PARTENZA"] = (
    df1.loc[null_mask, "AREOPORTO_PARTENZA"]
    .str.strip().str.lower()
    .map(iata_city)
)

print(f"Null left: {df1['CITTA_PARTENZA'].isnull().sum()}")
print(f"Unique values: {df1['CITTA_PARTENZA'].nunique()}")

print(df1['CITTA_PARTENZA'].value_counts(dropna=False).to_string())

Null left: 0
Unique values: 115
CITTA_PARTENZA
londra               1392
tirana                596
istanbul              417
dubai                 155
cairo                 125
manchester            125
casablanca            123
marrakech             104
new york              102
doha                  100
sharm el sheikh        88
tunis                  81
belgrado               80
abu dhabi              74
chisinau               57
newark                 57
jeddah                 51
san paolo              50
edinburgh              50
bristol                46
tel aviv yafo          44
amman                  44
addis ababa            41
birmingham             40
pechino                34
n.d.                   31
buenos aires           31
yerevan                31
shanghai               30
seoul                  28
giza                   27
nuova delhi            26
taipei                 25
algeri                 25
?                      23
nd                     23
kuwait city      

### Codice Paese Arrivo / Partenza

In [174]:
print("--- CODICE_PAESE_ARR ---")
print(f"Unique: {df1['CODICE_PAESE_ARR'].nunique()}  |  Null: {df1['CODICE_PAESE_ARR'].isnull().sum()}")
print(df1['CODICE_PAESE_ARR'].value_counts(dropna=False).to_string())
print()
print("--- CODICE_PAESE_PART ---")
print(f"Unique: {df1['CODICE_PAESE_PART'].nunique()}  |  Null: {df1['CODICE_PAESE_PART'].isnull().sum()}")
print(df1['CODICE_PAESE_PART'].value_counts(dropna=False).to_string())

--- CODICE_PAESE_ARR ---
Unique: 8  |  Null: 0
CODICE_PAESE_ARR
ita    4690
it      248
fra      25
gbr      18
esp      18
tur      17
deu      13
gb        1

--- CODICE_PAESE_PART ---
Unique: 67  |  Null: 54
CODICE_PAESE_PART
gbr        1626
alb         550
tur         407
egy         259
mar         251
are         234
usa         215
chn         116
qat         101
tun          83
srb          79
gb           78
sau          75
bra          66
mda          58
NaN          54
eth          46
jor          45
isr          43
al           35
arg          30
arm          29
ind          27
kor          26
dza          25
tr           24
unknown      24
twn          23
sgp          22
omn          22
kwt          22
mkd          17
eg           17
can          17
geo          17
mdv          16
             15
aze          15
nd           14
n.d.         13
bhr          13
//           13
irn          12
?            12
sen          12
mex          12
ae           12
mne          11
ken

In [175]:
# Cross-check: AREOPORTO_ARRIVO contains only Italian airports, so CODICE_PAESE_ARR
# must be Italy for every row. Any non-"ita" value is a data-entry error.
# We overwrite the entire column rather than mapping variants, since the ground
# truth (Italian arrival airport) is unambiguous.

n_wrong = (df1["CODICE_PAESE_ARR"] != "ita").sum()
print(f"Rows with non-'ita' value before fix: {n_wrong}")
if n_wrong > 0:
    print(df1.loc[df1["CODICE_PAESE_ARR"] != "ita", "CODICE_PAESE_ARR"].value_counts().to_string())

df1["CODICE_PAESE_ARR"] = "ita"

print(f"\nUnique values after fix: {df1['CODICE_PAESE_ARR'].nunique()}")
print(df1["CODICE_PAESE_ARR"].value_counts().to_string())

# CODICE_PAESE_PART: recover nulls using IATA → ISO-3 country code lookup.
iata_country_code = {
    "tia": "alb", "stn": "gbr", "lgw": "gbr", "lhr": "gbr", "lcy": "gbr",
    "ltn": "gbr", "bqh": "gbr", "fab": "gbr", "nht": "gbr", "lpl": "gbr",
    "ncl": "gbr", "edi": "gbr", "bhx": "gbr", "gla": "gbr", "bfs": "gbr",
    "lba": "gbr", "ema": "gbr", "brs": "gbr", "man": "gbr", "oxf": "gbr",
    "ist": "tur", "saw": "tur", "adb": "tur",
    "dxb": "are", "shj": "are", "auh": "are", "dwc": "are",
    "cai": "egy", "ssh": "egy", "hrg": "egy", "lxr": "egy", "rmf": "egy",
    "cmn": "mar", "rak": "mar", "tng": "mar", "fez": "mar", "aga": "mar", "rba": "mar",
    "doh": "qat",
    "jfk": "usa", "ewr": "usa", "mia": "usa", "atl": "usa", "dfw": "usa",
    "iad": "usa", "bos": "usa", "phl": "usa", "oak": "usa", "lax": "usa",
    "tun": "tun", "beg": "srb",
    "jed": "sau", "ruh": "sau", "dmm": "sau",
    "gru": "bra", "gig": "bra",
    "eze": "arg", "add": "eth",
    "mex": "mex", "cun": "mex",
    "pek": "chn", "pvg": "chn", "szx": "chn", "can": "chn", "wnz": "chn",
    "hgh": "chn", "nkg": "chn", "cgo": "chn", "tfu": "chn", "ckg": "chn",
    "rmo": "mda", "kiv": "mda",
    "hkg": "hkg", "mct": "omn", "mle": "mdv", "evn": "arm",
    "tbs": "geo", "kut": "geo", "icn": "kor", "sid": "cpv", "spx": "egy",
    "ika": "irn", "prn": "xkx", "mba": "ken", "znz": "tza",
    "tpe": "twn", "sin": "sgp", "tlv": "isr", "amm": "jor",
    "del": "ind", "atq": "ind", "kwi": "kwt",
    "yul": "can", "yyz": "can", "hnd": "jpn", "bey": "lbn", "alg": "dza",
    "gyd": "aze", "skp": "mkd", "tgd": "mne", "anu": "atg", "dss": "sen",
    "bah": "bhr", "lrm": "lby", "mji": "lby", "mbj": "jam", "mru": "mus",
    "tas": "uzb", "sll": "sle", "ala": "kaz", "isl": "pak",
}

null_mask = df1["CODICE_PAESE_PART"].isnull()
df1.loc[null_mask, "CODICE_PAESE_PART"] = (
    df1.loc[null_mask, "AREOPORTO_PARTENZA"]
    .str.strip().str.lower()
    .map(iata_country_code)
)

print(f"\nNull left CODICE_PAESE_PART: {df1['CODICE_PAESE_PART'].isnull().sum()}")
print(df1['CODICE_PAESE_PART'].value_counts(dropna=False).head(15).to_string())

Rows with non-'ita' value before fix: 340
CODICE_PAESE_ARR
it     248
fra     25
gbr     18
esp     18
tur     17
deu     13
gb       1

Unique values after fix: 1
CODICE_PAESE_ARR
ita    5030

Null left CODICE_PAESE_PART: 0
CODICE_PAESE_PART
gbr    1648
alb     559
tur     411
egy     262
mar     251
are     236
usa     219
chn     118
qat     102
tun      83
srb      80
gb       78
sau      75
bra      67
mda      58


### Paese Arrivo / Partenza

In [176]:
print("--- PAESE_ARR ---")
print(f"Unique: {df1['PAESE_ARR'].nunique()}  |  Null: {df1['PAESE_ARR'].isnull().sum()}")
print(df1['PAESE_ARR'].value_counts(dropna=False).to_string())
print()
print("--- PAESE_PART ---")
print(f"Unique: {df1['PAESE_PART'].nunique()}  |  Null: {df1['PAESE_PART'].isnull().sum()}")
print(df1['PAESE_PART'].value_counts(dropna=False).to_string())

--- PAESE_ARR ---
Unique: 1  |  Null: 0
PAESE_ARR
italia    5030

--- PAESE_PART ---
Unique: 57  |  Null: 74
PAESE_PART
regno unito              1694
albania                   587
turchia                   421
egitto                    267
marocco                   253
emirati arabi uniti       243
stati uniti               220
cina                      117
qatar                     101
tunisia                    81
serbia                     80
arabia saudita             76
NaN                        74
brasile                    63
moldavia                   56
giordania                  45
etiopia                    44
israele                    42
argentina                  30
india                      29
armenia                    28
corea del sud              27
algeria                    24
taiwan                     22
kuwait                     22
singapore                  21
oman                       21
nd                         21
//                         21
?         

In [177]:
# PAESE_ARR: always Italy — no action needed.
# PAESE_PART: recover nulls using ISO-3 code already filled in CODICE_PAESE_PART.
codice_paese_map = {
    "alb": "albania",          "gbr": "regno unito",      "tur": "turchia",
    "are": "emirati arabi",    "egy": "egitto",            "mar": "marocco",
    "qat": "qatar",            "usa": "stati uniti",       "tun": "tunisia",
    "srb": "serbia",           "sau": "arabia saudita",    "bra": "brasile",
    "arg": "argentina",        "eth": "etiopia",           "mex": "messico",
    "chn": "cina",             "mda": "moldavia",          "hkg": "hong kong",
    "omn": "oman",             "mdv": "maldive",           "arm": "armenia",
    "geo": "georgia",          "kor": "corea del sud",     "cpv": "capo verde",
    "irn": "iran",             "xkx": "kosovo",            "ken": "kenya",
    "tza": "tanzania",         "twn": "taiwan",            "sgp": "singapore",
    "isr": "israele",          "jor": "giordania",         "ind": "india",
    "kwt": "kuwait",           "can": "canada",            "jpn": "giappone",
    "lbn": "libano",           "dza": "algeria",           "aze": "azerbaigian",
    "mkd": "macedonia",        "mne": "montenegro",        "atg": "antigua",
    "sen": "senegal",          "bhr": "bahrein",           "lby": "libia",
    "jam": "giamaica",         "mus": "mauritius",         "uzb": "uzbekistan",
    "sle": "sierra leone",     "kaz": "kazakistan",        "pak": "pakistan",
}

null_mask = df1["PAESE_PART"].isnull()
df1.loc[null_mask, "PAESE_PART"] = (
    df1.loc[null_mask, "CODICE_PAESE_PART"]
    .str.strip().str.lower()
    .map(codice_paese_map)
)

print(f"Null left PAESE_PART: {df1['PAESE_PART'].isnull().sum()}")
print(df1['PAESE_PART'].value_counts(dropna=False).head(15).to_string())

Null left PAESE_PART: 4
PAESE_PART
regno unito            1714
albania                 595
turchia                 429
egitto                  272
marocco                 254
emirati arabi uniti     243
stati uniti             225
cina                    119
qatar                   102
tunisia                  82
serbia                   81
arabia saudita           76
brasile                  64
moldavia                 56
giordania                46


### Zona

In [178]:
print(f"Unique values: {df1['ZONA'].nunique()}")
print(f"Null values: {df1['ZONA'].isnull().sum()}")
print()
print(df1['ZONA'].value_counts(dropna=False).sort_index().to_string())

Unique values: 12
Null values: 0

ZONA
-1       3
1      177
10       1
2     1821
4      559
5     1457
6      210
7      143
8      575
9       78
99       1
??       5


In [179]:
# -1 is not a valid zone — treat it as missing before casting to integer.
df1["ZONA"] = df1["ZONA"].replace("-1", np.nan)
df1["ZONA"] = pd.to_numeric(df1["ZONA"], errors="coerce")
df1.loc[df1["ZONA"] == -1, "ZONA"] = np.nan   # catch numeric -1 if already parsed
df1["ZONA"] = df1["ZONA"].astype(pd.Int64Dtype())

print(f"Null left (were -1): {df1['ZONA'].isnull().sum()}")
print(f"dtype: {df1['ZONA'].dtype}")
print(df1["ZONA"].value_counts(dropna=False).sort_index().to_string())

Null left (were -1): 8
dtype: Int64
ZONA
1        177
2       1821
4        559
5       1457
6        210
7        143
8        575
9         78
10         1
99         1
<NA>       8


### TOT

In [180]:
print(f"Unique values: {df1['TOT'].nunique()}")
print(f"Null values: {df1['TOT'].isnull().sum()}")
print()
print(df1['TOT'].value_counts(dropna=False).head(20).to_string())

Unique values: 348
Null values: 0

TOT
1       2351
2        174
3        148
4        113
7         84
5         83
8         75
6         68
10        50
9         47
0         41
?         40
15        40
n.d.      35
16        34
12        32
11        32
14        31
18        28
13        27


In [181]:
# Cast to integer.
df1["TOT"] = pd.to_numeric(df1["TOT"], errors="coerce").astype(pd.Int64Dtype())

print(f"Null left: {df1['TOT'].isnull().sum()}")
print(f"dtype: {df1['TOT'].dtype}")
print(f"Min: {df1['TOT'].min()}  |  Max: {df1['TOT'].max()}  |  Mean: {df1['TOT'].mean():.1f}")

TypeError: cannot safely cast non-equivalent float64 to int64

### Motivo Allarme

In [ ]:
print(f"Unique values: {df1['MOTIVO_ALLARME'].nunique()}")
print(f"Null values: {df1['MOTIVO_ALLARME'].isnull().sum()}")
print()
print(df1['MOTIVO_ALLARME'].value_counts(dropna=False).to_string())

In [ ]:
# Standardise category labels.
motivo_map = {
    "sdi"      : "sdi",
    "nsis"     : "nsis",
    "sdi/nsis" : "sdi/nsis",
    "tsc"      : "tsc",
    "interpol" : "interpol",
    "bcs"      : "bcs",
    "manuale"  : "manuale",
}

df1["MOTIVO_ALLARME"] = df1["MOTIVO_ALLARME"].map(lambda x: motivo_map.get(str(x).strip(), np.nan) if pd.notna(x) else np.nan)

print(f"Null left: {df1['MOTIVO_ALLARME'].isnull().sum()}")
print(df1["MOTIVO_ALLARME"].value_counts(dropna=False).to_string())

### Note Operatore

In [ ]:
print(f"Unique values: {df1['note_operatore'].nunique()}")
print(f"Null values: {df1['note_operatore'].isnull().sum()}  ({df1['note_operatore'].isnull().mean()*100:.1f}%)")
print()
print(df1['note_operatore'].value_counts(dropna=False).head(15).to_string())

### Flag Rischio

In [ ]:
print(f"Unique values: {df1['flag_rischio'].nunique()}")
print(f"Null values: {df1['flag_rischio'].isnull().sum()}  ({df1['flag_rischio'].isnull().mean()*100:.1f}%)")
print()
print(df1['flag_rischio'].value_counts(dropna=False).to_string())

### Colonne Duplicate

In [ ]:
# The duplicate columns may use different formatting (case, punctuation, encoding)
# from their cleaned counterparts. Normalise both sides before comparing.

def norm_str(s):
    """Lowercase, strip whitespace and trailing dots; treat null-like strings as empty."""
    if pd.isna(s):
        return ""
    s = str(s).strip().lower().rstrip(".")
    return "" if s in {"nan", "none", "n/a", "nd", "n.d.", "<na>"} else s

str_checks = [
    ("Paese Partenza",   "PAESE_PART"),
    ("CODICE PAESE ARR", "CODICE_PAESE_ARR"),
    ("paese%arr",        "PAESE_ARR"),
]

for dup, orig in str_checks:
    nd = df1[dup].map(norm_str)
    no = df1[orig].map(norm_str)
    n_mismatch = (nd != no).sum()
    print(f"'{dup}' ~= '{orig}':  {n_mismatch} mismatches out of {len(df1)}")
    if n_mismatch > 0 and n_mismatch <= 10:
        print(df1.loc[nd != no, [dup, orig]].to_string())
    elif n_mismatch > 10:
        print("  sample differences:")
        print(df1.loc[nd != no, [dup, orig]].head(5).to_string())

# Numeric check: tot voli (str) vs TOT (Int64)
n_mismatch_tot = (
    pd.to_numeric(df1["tot voli"], errors="coerce") != df1["TOT"].astype(float)
).sum()
print(f"\n'tot voli' ~= 'TOT':  {n_mismatch_tot} mismatches")

# Integer check: 3zona (int64) vs ZONA (Int64, -1 → NaN)
# For rows where ZONA is NA (was -1), '3zona' should have -1 → skip those
valid = df1["ZONA"].notna()
n_mismatch_zona = (df1.loc[valid, "3zona"] != df1.loc[valid, "ZONA"].astype(int)).sum()
na_zona = (~valid).sum()
print(f"'3zona' ~= 'ZONA':  {n_mismatch_zona} mismatches on valid rows  ({na_zona} rows had -1 → NaN)")

In [ ]:
# Drop the duplicate columns.
cols_to_drop = ["Paese Partenza", "CODICE PAESE ARR", "3zona", "paese%arr", "tot voli"]
df1.drop(columns=cols_to_drop, inplace=True)

print(f"Remaining columns ({df1.shape[1]}):")
print(df1.columns.tolist())

# Tipologia Viaggiatore Dataset

First general look at the dataset (df2)

In [ ]:
df2.shape

(5095, 33)

In [ ]:
df2.head(5)

,NAZIONALITA,AREOPORTO_ARRIVO,AREOPORTO_PARTENZA,ANNO_PARTENZA,MESE_PARTENZA,GIORNO_PARTENZA,DATA_PARTENZA,DESCR_AEREOPORTO_ARR,DESCR_AEREOPORTO_PART,CITTA_ARR,...,COMPAGNIA_AEREA,NUMERO_VOLO,ESITO_CONTROLLO,note_operatore,codice_rischio,Tipo Documento,FASCIA ETA,3nazionalita,compagnia%aerea,num volo
0,ALB,NAP,DUR,2024,02,13,2024-02-13 07:30:00,Napoli Capodichino,King Shaka International,Napoli,...,Fly Dubai,FZ1681,RESPINTO,NaN,NaN,Passaporto,N.D.,ALB,Fly Dubai,FZ1681
1,NaN,FCO,JFK,2024,01,22,2024-01-22 16:35:00,Fiumicino,John F Kennedy International,Roma,...,ITA Airways,AZ0609,NaN,NaN,NaN,Carta d'identità,18-30,ALB,ITA Airways,AZ0609
2,ALB,TSF,TIA,2024,02,4,2024-02-04 20:10:00,Treviso-Sant'Angelo,Rinas Mother Teresa,Treviso,...,Ryanair DAC,FR8400,SEGNALATO,NaN,NaN,N.D.,31-45,ALB,Ryanair DAC,FR8400
3,AFG,FCO,IST,2024,01,25,2024-01-25 13:05:00,Fiumicino,Havalimani,Roma,...,Turkish Airlines,TK1865,NaN,NaN,NaN,N.D.,61+,AFG,Turkish Airlines,TK1865
4,ALB,BGY,MLE,2024,02,13,FEB 13 2024,Orio al Serio,Male International,Bergamo,...,Fly Dubai,FZ1571,SEGNALATO,NaN,NaN,Permesso di soggiorno,46-60,ALB,Fly Dubai,FZ1571


In [ ]:
df2.info()

<class 'pandas.DataFrame'>
RangeIndex: 5095 entries, 0 to 5094
Data columns (total 33 columns):
 #   Column                 Non-Null Count  Dtype
---  ------                 --------------  -----
 0   NAZIONALITA            4979 non-null   str  
 1   AREOPORTO_ARRIVO       5095 non-null   str  
 2   AREOPORTO_PARTENZA     5095 non-null   str  
 3   ANNO_PARTENZA          5095 non-null   str  
 4   MESE_PARTENZA          5095 non-null   str  
 5   GIORNO_PARTENZA        5095 non-null   int64
 6   DATA_PARTENZA          5095 non-null   str  
 7   DESCR_AEREOPORTO_ARR   5095 non-null   str  
 8   DESCR_AEREOPORTO_PART  5095 non-null   str  
 9   CITTA_ARR              5095 non-null   str  
 10  CITTA_PARTENZA         5095 non-null   str  
 11  CODICE_PAESE_ARR       5095 non-null   str  
 12  CODICE_PAESE_PART      5095 non-null   str  
 13  PAESE_ARR              5095 non-null   str  
 14  PAESE_PART             5095 non-null   str  
 15  ZONA                   5095 non-null   str  
 16 